xla/service/float8_fnuz_ir_emitter.cc

```diff
@@ -0,0 +1,646 @@
+// ExtractMantissa(value) = 0b0000000110111111
+//                                    ^- 2^(-17) is here at bit 7.
+// result = LastMantissaBit(F16, 0.00002664, F8E5M2FNUZ, b) = 7
+//
+// Example 2:
+// input_type  = BF16
+// output_type = F8E4M3FNUZ
+// value = 247.0
+//       = 0x1.EEp7
+//       = 0x4377
+//       = 0b0|10000110|1110111
+//       = 0b1.1110111 * 2^7
+// Given the input and output is a normal number, we are looking for the bit
+// the corresponds to the third bit of the mantissa.
+// ExtractMantissa(value) = 0b0000000011110111
+//                                       ^- third mantissa bit is at bit 4.
+// result = LastMantissaBit(BF16, 247.0, F8E4M3FNUZ, b) = 4
+llvm::Value* LastMantissaBit(PrimitiveType input_type, llvm::Value* value,
+                             PrimitiveType output_type, llvm::IRBuilder<>* b) {
+  const int src_mantissa_bits = SignificandWidth(input_type) - 1;
+  const int dest_mantissa_bits = SignificandWidth(output_type) - 1;
+  llvm::Type* int_type = b->getIntNTy(BitWidth(input_type));
+
+  llvm::Value* exponent = ExtractExponent(input_type, value, b);
+
+  // The index when the input/output is normal.
+  llvm::Value* last_bit_index =
+      llvm::ConstantInt::get(int_type, src_mantissa_bits - dest_mantissa_bits);
+
+  // Increase the index if the output will be denormal given the exponent.
+  llvm::Value* denormal_shift = b->CreateSub(
+      llvm::ConstantInt::get(int_type, UnderflowExponent(output_type) - 1),
+      exponent);
+  denormal_shift = Max(int_type, denormal_shift, 0, b);
+  last_bit_index = b->CreateAdd(last_bit_index, denormal_shift);
+
+  // The log_2(x) of the smallest denorm value. This gives us the exponent n
+  // that produces that number. This corresponds to the encoding with a single
+  // bit set in the last significant bit (0b0000...01).
+  const int input_log_minimum =
+      UnderflowExponent(input_type) - SignificandWidth(input_type);
+  const int output_log_minimum =
+      UnderflowExponent(output_type) - SignificandWidth(output_type);
+
+  // Alternatively, the input might be a denorm. This directly computes the
+  // last mantissa bit when the input is a denorm.  Suppose we have 2.664E-5
+  // (0x1.BFp-16) as the input number. The bit encoding for this in F16 is:
+  //   S|EEEEE|MMMMMMMMMM
+  // 0b0|00000|0110111111 = 2^(-14) * 0b0.0110111111
+  //
+  // To cast this to F8E5M2FNUZ, we would find the smallest denorm encoding and
+  // find the corresponding bit in the input.
+  //   S|MMMMM|MM
+  // 0b0|00000|01 = 2^(-15) * 0b0.01
+  // We can see the "last" bit of an output denorm represents 2^(-17), so we
+  // find that corresponding bit in the input. In this F16 example it is
+  // highlighted here:
+  //   S|EEEEE|MMMMMMMMMM
+  // 0b0|00000|0110111111
+  //             ^-- last mantissa bit
+  llvm::Value* denorm_last_mantissa_bit =
+      llvm::ConstantInt::get(int_type, output_log_minimum - input_log_minimum);
```

Comment:
I think this still has the same issue I described with the hypothetical E5M2a type in [my previous comment](https://github.com/openxla/xla/pull/3200#discussion_r1236178397). Going from F16 to E5M2a would cause `denorm_last_mantissa_bit` to be 6, but the function should return 7.

I think this logic works as long as the exponent bias of the output type is at most 1 greater than the exponent bias of the input type. This is true for all existing floating-point types where the input type is wider than the output type, and the only such case where the output type has a greater exopnent bias is when going from F16 to E5M2 FNUZ, in which case E5M2 FNUZ's exponent bias is only greater by 1.

If you comment this restriction and DCHECK it, I'm fine with the current logic. Alternatively, you can special case the case where you go from F16 to E5M2 FNUZ.

Reply:
I think a value of 6 is the correct answer for the hypothetical `E5M2a` type.
the smallest positive non-zero value is `0b0|00000|01`, which is `0b0.01 * 2^(-16)` which equals `2^(-18)`.
The `2^(-18)` bit in the F16 denorm is bit 6.

This can be seen by taking `2^(-18)` and storing it in `F16`. The hex for this `F16` value is `0x0040` in hex, which does indeed have bit 6 set and corresponds to the "last mantissa bit" of our output type.

Reply:
Okay, I get the case you mean when a denorm maps to a normal number in the output type. I've added a check for this to reject any future hypothetical types that break this assumption.